File: Python/DynamicArray/dynamic_array.py
# create dynamic array the can add/delete values to end, beginning, or specific location in array. Do not use any methods to help. Do everything in place without creating new arrays.

class DynamicArray:
  def __init__(self, arr = []):
    self.arr = arr
  # append value to end of array
  def addToBack(self, value):
    self.arr[len(self.arr):] = [value]
    return self
  # prepend value to beginning of array
  def addToFront(self, value):
    self.arr[:0] = [value]
    return self
  # remove back of array
  def removeBack(self):
    self.arr = self.arr[:(len(self.arr) - 1)]
    return self
  # remove front of array
  def removeFront(self):
    self.arr = self.arr[1:len(self.arr)]
    return self
  # insert value at specific location
  def insertValue(self, value=None, idx=None):
    if idx == None:
      return self
    if idx > len(self.arr) or value == None:
      return self
    elif idx == len(self.arr):
      self.addToBack(value)
      return self
    else:
      self.arr = self.arr[:idx] + [value] + self.arr[idx:]
      return self
  # remove value at idx recieved
  def removeIndex(self, idx=None):
    if idx > len(self.arr) or idx == None:
      return self
    elif idx == (len(self.arr) - 1):
      self.removeBack()
      return self
    elif idx == 0:
      self.removeFront()
      return self
    else:
      self.arr = self.arr[:idx] + self.arr[idx + 1:]
      return self
  # remove first value that matches value given
  def removeValue(self, value=None):
    if value == None:
      return self
    for i in range(0, len(self.arr)):
      if self.arr[i] == value:
        self.removeIndex(i)
        break
    return self

File: JavaScript/DynamicArray/DynamicArray.js
/*
create dynamic array the can add / delete values to end, beginning, or specific location in array.Do not use any methods to help.Do everything in place without creating new arrays.
*/

class DynamicArray {
  constructor(arr = []) {
    this.arr = arr;
  }
  // add value to end of array
  addToBack = value => {
    this.arr[this.arr.length] = value;
    return this;
  }
  // add value to front of array
  addToFront = value => {
    for(let i = this.arr.length - 1; i >= 0; i--) {
      this.arr[i + 1] = this.arr[i];
    }
    this.arr[0] = value;
    return this;
  }
  // Remove last element from array
  removeBack = _=> {
    this.arr.length = this.arr.length - 1;
    return this;
  }
  // remove first element from array
  removeFront = _=> {
    for (let i = 0; i < this.arr.length; i ++) {
      this.arr[i] = this.arr[i + 1];
    }
    this.removeBack();
    return this;
  }
  // insert value at specific location in array
  insertAt = (value, idx=null) => {
    if(value === undefined || idx === null || idx > this.arr.length) {
      return this;
    } else if (idx === this.arr.length) {
      this.addToBack(value);
    } else if (idx === 0) {
      this.addToFront(value);
      return this;
    } else {
      this.arr[this.arr.length] = value;
      for (let i = this.arr.length - 2; i >= 0; i--) {
        if (i >= idx) {
          let temp = this.arr[i];
          this.arr[i] = this.arr[i + 1];
          this.arr[i + 1] = temp;
        }
      }
      return this;
    }
  }
  // remove value at index
  removeIndex(idx=null){
    if(idx === null || idx >= this.arr.length || idx < 0){
      return this;
    } else if (idx === 0) {
      this.removeFront();
    } else if (idx === this.arr.length - 1) {
      this.removeBack()
    } else {
      for (let i = 0; i < this.arr.length; i++) {
        if (i >= idx) {
          this.arr[i] = this.arr[i + 1];
        }
      }
      this.removeBack();
      return this;
    }
  }
  // remove first value that matches value given
  removeValue = (value) => {
    if (value === undefined) {
      return this;
    } else {
      for (let i = 0; i < this.arr.length; i++) {
        if (this.arr[i] === value) {
          this.removeIndex(i);
          break;
        }
      }
    }
    return this;
  }
}

// const myArr = new DynamicArray([1,2,3,4,5,6,7]);

// myArr.addToBack(7);
// console.log(myArr.arr);

module.exports = DynamicArray;